Linux: Try a CosmicText based Text System

h3mosphere opened the pull request:
This is a rebase of @gabydds text_system updates. with some small cleanups.

Currently cannot test this as build is not working in linux. Im just putting it up here before I forget about it.

gabydd committed:
flake

gabydd committed:
initial text system

gabydd committed:
some text gets rendered

gabydd committed:
add some comments

gabydd committed:
a little better

gabydd committed:
correct placement

gabydd committed:
smaller gutter, this is not correct but it works

gabydd committed:
looks a bit better, and some basic font fallback

h3mosphere committed:
remove some commented mac specific code, unused vars
I have left some println!'s commented, just in case gabydd needs to look at this

h3mosphere committed:
return all font names

h3mosphere committed:
revert todo changes for clean text-system-cleanup rebase

h3mosphere committed:
Revert "flake"

This reverts commit 4bd702caaaece03d48bb119e4eed047e54d8042e.

h3mosphere committed:
add todo!(linux) comments back in. Most of gabydd's comments were him indicating problematic parts of the code, so these are probably all todo's

h3mosphere committed:
Merge branch 'main' into text-system-cleanup-5

h3mosphere commented:
Updated to main. Hopefully good to go.

h3mosphere commented:
Depends on #7555

gabydd commented:
Did #7555 actually fix the rendering for you when I merged it and master too my branch with input it only half rendered and wouldn't expect input hopefully that was just a bad merge

ForLoveOfCats commented:
Does this need to pull in `swash` directly? It looks like it only uses the re-exports provided by Cosmic Text?

gabydd commented:
Doesn't need swash directly, that was just a historical artifact of how I was doing metrics conversion at one point

h3mosphere commented:
Hey,

I cleaned this up a little for inclusion @kvark 's branch, as per discord discussion, but then it landed in main. It's raw, just really what @gabydd came to after an inspired hacking session. It gets zed to the point of running, a minimal baseline with which to hack further.

I probably should have continued on with the blade branch, and cleaned this code up, but got busy trying to get it work with main.

gabydd commented:
No worries, of the top of my head here is a little Todo list:

- [] figure out why the scroll track is extra long(check against mac cause maybe this is just the regular behavior)
- [] figure out why the seperator between the gutter and editor disapears
- [] font loading
  - [] look at getting system font fallback to work, hopefully can do this with font-db if not maybe use font-kit or look what they do
   - [] figure out if we actually need load_family/should change how we query from font-db, right now when first querying for a font id we almost always fallback to creating a new one because the font given by font id is not in the map
  - [] test if font-loading is slow, seems to take a bit of time but maybe that is something else
- [] I would love some more eyes on render_line, I don't think we need to use cosmic-texts Buffer here because we already do caching(we are double caching glyph rasterization actually both gpui and cosmic text have a cache), but its definetly possible I am not interfacing with cosmic text properly and I did get a crash at one point but wasn't able to reproduce it
- [] optionally double check against font-kit that we are giving the correct metrics

gabydd commented on crates/gpui/src/platform/linux/text_system.rs:
```diff
@@ -49,66 +61,129 @@ impl Default for LinuxTextSystem {
+        FontMetrics {
+            units_per_em: metrics.units_per_em as u32,
+            ascent: metrics.ascent,
+            descent: metrics.descent - 1000.0, // todo!(linux) extremely hacky but this means text doesn't get cut off, something to do with gutter padding in the editor
```
I mentioned this on the other pr but :

instead of subtracting 1000 I think this should just be -metrics.descent like in https://github.com/pop-os/cosmic-text/blob/990d66ed41df4fae5db23ec10c9d6c0e95854449/src/shape.rs#L103

h3mosphere replied:
Ah, I forgot about this.

@mikayla-maki apologies I force pushed over your fmt,

gabydd replied:
yep the subtract makes wrap overflow: ![image](https://github.com/zed-industries/zed/assets/82682503/4c7e697b-f579-44fe-96f5-fa3ae2b51cb1)
and minus looks good: ![image](https://github.com/zed-industries/zed/assets/82682503/dfedb9dc-02bc-4da0-8743-36e79dde0f47)

h3mosphere committed:
-metrics.descent in FontMetrics returned

h3mosphere committed:
fmt

h3mosphere committed:
spelling

mikayla-maki committed:
Adjust the comments a little bit

mikayla-maki committed:
Get text showing up

mikayla-maki commented:
With this PR, we can now render the hello world example!

![Screenshot_from_2024-02-08_22-43-31](https://github.com/zed-industries/zed/assets/2280405/612548f9-a494-4081-8f72-375aa47eb625)

mikayla-maki commented:
> _What needs to be done:_

Thank you for the comprehensive list! I think we can leave these for PRs as we go :)
